nsqadmin/static/js/app_state.js

```diff
@@ -30,8 +31,15 @@ var AppState = Backbone.Model.extend({
         this.set('graph_interval', interval);
     },
 
-    url: function(url) {
-        return '/api' + url;
+    basePath: function(p) {
+        // if base path is / then don't prefix
+        var bp = this.get('BASE_PATH') == '/' ? '' : this.get('BASE_PATH');
+        // remove trailing /, but guarantee at least /
+        return (bp + p).replace(/\/$/, '') || '/';
```

Comment:
this removes the trailing slash from `p` ... did you intend to remove the trailing slash from `bp`?

Reply:
There's some unexpected mess and complexity in this approach ... I think it can be simplified to "just remove trailing slash from basepath".

If the base path is `"/"` then we want `""`
If the base path is `"/admin/"` then we want `"/admin"`

Then, you can append anything to that, e.g. `"/topics"`, or just `"/"` to get the root URL path.

Reply:
I guess this function is pretty much the only problem here, the rest looks good.

In Go land, `path.Join()` takes care of everything after the leading `/`.

So I guess this is the only place where you might have to worry about a double-slash between `BASE_PATH` and `p`.

This logic seems a bit much for what should still be a trivial function. I think my initial suggestion here was appropriate, just remove trailing slash from `BASE_PATH`.

Reply:
... e.g.

```javascript
return this.get('BASE_PATH').replace(/\/$/, '') + p;
```

but since this is the only place `BASE_PATH` is used, you could do the regex just once, in `index.html`:

```js
var BASE_PATH = {{basePath ""}}.replace(/\/$/, '');
```

and then here it's just

```js
return this.get('BASE_PATH') + p;
```

Reply:
The case this is addressing is, e.g. where `BASE_PATH` is `/base` and in one of our handlebars templates we have `{{basePath "/"}}`. What we _want_ is `/base` _not_ `/base/`, but if `BASE_PATH` is `/` we want `/` not `//`.

Basically, the answer to your original question is that I intended to remove any potential trailing slash from the joined path.

Also, FWIW, I think `BASE_PATH` the global var that's passed into `AppState` should always represent the normalized value of whatever was passed to `nsqadmin --base-path`.

If you can come up with an easier way to do this, I'm all ears.

Reply:
Look at the go function `bp()` in:

```go
router.Handle("GET", bp("/"), http_api.Decorate(s.indexHandler, log))
```

- that code actually gets you `/base/`, so if you wanted `/base`, then it's wrong. Personally, I think `/base/` is good.

---

Notice that your remove-trailing-slash regex _only ever removes a slash from `p`_ because the characters at the end of `(bp + p)` are always from `p`. So the only value of `p` this could be useful for is `p == "/"` in which case you could instead handle it like `if (p == "/") return this.get('BASE_PATH')`. (Recall that `bp` is the user-supplied base path and `p` is the app-meaningful part from our code.)

---

There is a super easy way. Normalize the BasePath to start with a `/`, and then to not end with a `/`, including turning `"/"` into `""`. Now a plain string concatenation works everywhere - in Go, in Javascript, everywhere. No conditional logic, no path.Join(), nothing special. You just have to accept that the root page path is `/base/`, and all the conditionals go away.

Reply:
I swear I tested `path.Join()` to see how it would join a slash, but I must have remembered wrong, and I was wrong here, the result of `bp("/")` would indeed be just `"/base"`.

I just thought of this whole thing in an apparently opposite way from how you. Your original way was probably quite simple, and the way I originally thought of it was also quite simple. But different enough that I didn't understand the purpose of some of your logic.

Anyway, whatever works, works.

Reply:
Why would we choose a path for a new page that ends in `/`, only for our code to remove that `/` here?

There still are special cases for `/` in both variables - but I think a formulation like this is more clear / less tricky:

```js
var bp = this.get('BASE_PATH');
if (bp == '/')
  return p;
if (p == '/')
  return bp;
return bp + p;
```

Reply:
fwiw, I think Backbone can handle trailing slash:

> Trailing slashes are treated as part of the URL, and (correctly) treated as a unique route when accessed. docs and docs/ will fire different callbacks.

http://backbonejs.org/#Router-routes

I'm a bit more surprised that it can handle the empty string when the leading slash is removed from `this.route(bp('/'), 'topics');` but I think that would happen either way. (It's fun to think through this case for the default base-path: that `bp()` calls this `basePath()`, `BASE_PATH` is "/" so prefix with empty string, remove "/", that results in empty string so return "/" instead, finally `bp()` removes first slash, resulting in empty string.)

Reply:
Yes, I know that Backbone can support it, I didn't mean to imply otherwise. I just think it's inconsistent with all the other routes that are _not_ currently configured to allow it (although, again, we never link to them without a trailing slash).

This feels like a good baseline so I think we should land this. If you're so inclined, I look forward to your follow up PR where I encourage you to explore this and any other rabbit hole that piques your interest 😉. I am totally willing to land something that demonstrates a simpler overall approach.

Reply:
yes, let's merge this